doctest/extensions/doctest_mpi.h

```diff
@@ -2,80 +2,104 @@
 
 #ifdef DOCTEST_CONFIG_IMPLEMENT
 
-#include "doctest/doctest.h"
+#include "doctest/extensions/mpi_sub_comm.h"
 #include "mpi_reporter.h"
-
-#else
-
-#include "mpi.h"
-#include <numeric>
-#include <vector>
-#include "doctest/doctest.h"
-#include <cassert>
-#include <string>
+#include <unordered_map>
 
 namespace doctest {
 
-inline
-int mpi_world_nb_procs() {
-  int n;
-  MPI_Comm_size(MPI_COMM_WORLD, &n);
-  return n;
+// Each time a MPI_TEST_CASE is executed on N procs,
+// we need a sub-communicator of N procs to execute it.
+// It is then registered here and can be re-used
+// by other tests that requires a sub-comm of the same size
+std::unordered_map<int,mpi_sub_comm> sub_comms_by_size;
+
+// Record if at least one MPI_TEST_CASE was registered "skipped"
+// because there is not enought procs to execute it
+int nb_test_cases_skipped_insufficient_procs = 0;
+
+// Can be safely called before MPI_Init()
+//   This is needed for MPI_TEST_CASE because we use doctest::skip()
+//   to prevent execution of tests where there is not enough procs,
+//   but doctest::skip() is called during test registration, that is, before main(), and hence before MPI_Init()
+int
+mpi_comm_world_size() {
+  #if defined(OPEN_MPI)
+    const char* size_str = std::getenv("OMPI_COMM_WORLD_SIZE");
+  #elif defined(I_MPI_VERSION) || defined(MPI_VERSION)
+    const char* size_str = std::getenv("PMI_SIZE"); // see https://community.intel.com/t5/Intel-oneAPI-HPC-Toolkit/Environment-variables-defined-by-intel-mpirun/td-p/1096703
+  #else
+    #error "Unknown MPI implementation"
+  #endif
+  if (size_str==nullptr) return 1; // not launched with mpirun/mpiexec, so assume only one process
+  return std::stoi(size_str);
```

Comment:
maybe add a check here to make sure it is > 0? Just to be safe

Reply:
On this suggestion, I have done the following : we record the value in a global variable, then test it against `MPI_Comm_size(MPI_COMM_WORLD)` during `doctest::mpi_init_thread`. So we should have an exception if this brittle `mpi_comm_world_size  `function has a problem (I don't like the function either... but don't know of any other way...)